### PR TITLE
Indicate missing support for unique network flows

### PR DESCRIPTION
**Pull request by anamud**
This PR indicates missing support in rmw_connextdds for unique network flows for publishers and subscribers in communicating nodes. Such indication is required for graceful error handling in the ROS2 RMW and above layers.

Associated pull-requests:
- [design](https://github.com/ros2/design/pull/304)
- [examples](https://github.com/ros2/examples/pull/296)
- [rclcpp](https://github.com/ros2/rclcpp/pull/1496)
- [rcl](https://github.com/ros2/rcl/pull/880)
- [rmw](https://github.com/ros2/rmw/pull/294)
- [rmw_implementation](https://github.com/ros2/rmw_implementation/pull/170)
- [rmw_fastrtps](https://github.com/ros2/rmw_fastrtps/pull/502)
- [rmw_cyclonedds](https://github.com/ros2/rmw_cyclonedds/pull/282)
- [rmw_connext](https://github.com/ros2/rmw_connext/pull/484) (NB: This repository is soon to be deprecated)
- [Fast DDS](https://github.com/eProsima/Fast-DDS/pull/1682)

Initial contributions stem from Ericsson and eProsima.

Signed-off-by: Ananya Muddukrishna <ananya.x.muddukrishna@ericsson.com>

### CODE REVIEW

File: rmw_connextdds_common/src/common/rmw_publication.cpp
Comment (asorbini) [inline rmw_connextdds_common/src/common/rmw_publication.cpp:148]: As you probably saw, this implementation defines a series of "feature flags" in `static_config.hpp` to be able to support multiple versions of ROS2 from the same branch. This is to avoid having to create branches for different versions, which we'll probably end up doing starting with Galactic, but would have been a pain to maintain up until now, while the RMW was in very active development.

For now, I'd like to continue to adhere with this pattern so that the `master` branch remains buildable with Foxy, so I suggest introducing a new flag in `static_config.hpp` to make sure field `require_unique_network_flow_endpoints` is only accessed in the appropriate ROS2 version. Something like:

```cpp
#ifndef RMW_CONNEXT_HAVE_UNIQUE_NETWORK_FLOWS
#define RMW_CONNEXT_HAVE_UNIQUE_NETWORK_FLOWS \
  (RMW_CONNEXT_RELEASE > RMW_CONNEXT_RELEASE_FOXY)
#endif /* RMW_CONNEXT_HAVE_UNIQUE_NETWORK_FLOWS */
```

Then wrap the new code in an `#if RMW_CONNEXT_HAVE_UNIQUE_NETWORK_FLOWS` (also in `rmw_subscription.cpp`).

File: rmw_connextdds_common/src/common/rmw_get_network_flow_endpoints.cpp
Comment (asorbini) [inline rmw_connextdds_common/src/common/rmw_get_network_flow_endpoints.cpp:49]: Package `rmw_connextdds_common` provides the implementation of core functions shared by both `rmw_connextdds` and `rmw_connextddsmicro`.

In order to make the Windows linker happy, each RMW package must define all "public API symbols" in their own source files so that they may available in their own shared libs (instead of the ones generated by `rmw_connextdds_common`). `rmw_connextdds` does this in [rmw_api_impl_ndds.cpp](https://github.com/ros2/rmw_connextdds/blob/master/rmw_connextdds/src/rmw_api_impl_ndds.cpp), while you'll find the same for `rmw_connextddsmicro` in [rmw_api_impl_rtime.cpp](https://github.com/ros2/rmw_connextdds/blob/master/rmw_connextddsmicro/src/rmw_api_impl_rtime.cpp).

The functions in these files are just proxies for functions [defined](https://github.com/ros2/rmw_connextdds/blob/master/rmw_connextdds_common/include/rmw_connextdds/rmw_api_impl.hpp) in `rmw_connextdds_common`, and implemented in various files in `rmw_connextdds_common/src/common`.

I suggest renaming this file to `rmw_networkflows.cpp`, to keep it a bit more generic and possibly more "forward-compatible" in case there are other things to add in the future related to "unique network flows".

The functions should be defined in `rmw_api_impl.hpp`, and I would rename them to:
- `rmw_api_connextdds_publisher_get_network_flow_endpoints`
- `rmw_api_connextdds_subscription_get_network_flow_endpoints`

The implementation is fine as is.

Both `rmw_api_impl_ndds.cpp` and `rmw_api_impl_rtime.cpp` should be modified to include the actual public functions, delegating to the implementation in `rmw_connextdds_common`, e.g.:

```cpp
rmw_ret_t
rmw_publisher_get_network_flow_endpoints(
  const rmw_publisher_t * publisher,
  rcutils_allocator_t * allocator,
  rmw_network_flow_endpoint_array_t * network_flow_endpoint_array)
{
  return rmw_api_connextdds_publisher_get_network_flow_endpoints(publisher, allocator, network_flow_endpoint_array);
}
```

**Review by asorbini — CHANGES_REQUESTED** [review]
Hi @anamud, thank you for your submission!

I left a couple of comments to make sure this code works with both `rmw_connextdds` and `rmw_connextddsmicro`, and with older versions of ROS2.

Once those are addressed, this PR will be ready to be merged, although it will need to be held off until [ros2/rmw#294](https://github.com/ros2/rmw/pull/294) is merged.

**Review by asorbini — APPROVED** [review]
Thank you for the updates @anamud!

The PR looks good now. I think two feature flags is a bit overkill, but I'm going to get rid of them altogether soon, so we can leave that as is.

At this point, I would only wait for [ros2/rmw#294](https://github.com/ros2/rmw/pull/294) to be merged.

**Comment by asorbini** [comment]
I created #15 to get rid of all "feature flags" in `master`. If you want, I think you could rebase your changes on `asorbini/rolling-no-ff` (or rebase them from `master`, once that PR is merged).

Apologies for the back and forth, but the recent streak of PRs made me realize getting rid of these flags had more priority than I thought.

**Comment by anamud** [comment]
> 
> 
> I created #15 to get rid of all "feature flags" in `master`. If you want, I think you could rebase your changes on `asorbini/rolling-no-ff` (or rebase them from `master`, once that PR is merged).
> 
> Apologies for the back and forth, but the recent streak of PRs made me realize getting rid of these flags had more priority than I thought.

@asorbini: Thanks for the review. I understand the rationale to get rid of the flags. I prefer to wait until the PR is merged before rebasing.

**Comment by asorbini** [comment]
#15 was merged, so you can rebase from `master` now and remove the feature flags

**Comment by anamud** [comment]
> 
> 
> #15 was merged, so you can rebase from `master` now and remove the feature flags

Thanks, @asorbini for the notice. I have now rebased to `master` and updated the PR. I removed conditional compilation altogether assuming a rolling release model.

File: rmw_connextdds_common/src/common/rmw_network_flow_endpoints.cpp
Comment (asorbini) [inline rmw_connextdds_common/src/common/rmw_network_flow_endpoints.cpp:19]: Looks like a feature flag made it through the refactoring :)

**Review by asorbini — CHANGES_REQUESTED** [review]

File: rmw_connextdds_common/src/common/rmw_network_flow_endpoints.cpp
Comment (anamud) [reply rmw_connextdds_common/src/common/rmw_network_flow_endpoints.cpp]: Good spotting! That one escaped my grep filter. I will push a new commit soon.

**Review by asorbini — APPROVED** [review]
LGTM.

It would be good to run CI with the other changed repositories, but I'm assuming that will happen as part of merging of (one of the) other PRs (let me know and I can also start some jobs).

In general, the PR can be merged once [ros2/rmw#294](https://github.com/ros2/rmw/pull/294) is merged.

**Review by wjwwood — APPROVED** [review]

**Comment by anamud** [comment]
@asorbini @wjwwood I removed a conflicting linkage [error](https://ci.ros2.org/job/ci_linux/14199/console#console-section-8) reported by the CI system. Please confirm that that [change](https://github.com/ros2/rmw_connextdds/pull/13/commits/28aa45844c5c081591eddff0c88feb2412a88244) is correct. 

I am unable to test this PR out properly due to lack of a licensed connextdds system at my side. Can you please suggest an alternative test method? 

Since the PRs related to unique network flows are now approved, I would appreciate it if you can start a CI job at your side.

**Comment by asorbini** [comment]
Hi @anamud, the change looks good to me, I think it's fine to have these function have C++ linkage, apologies for not spotting that in my earlier reviews.

About the licensing issue: I'm not a legal expert and this is not legal advice, but maybe this ROS 2 development work counts towards "research purposes" covered by the "non-commercial license" included in Debian package `rti-connext-dds-5.3.1`. If that is the case, than you could also submit a request for a permanent "reserach and non-commercial use" license from [this RTI webpage](https://www.rti.com/free-trial) and have access to Connext for other platforms too. This would also give you access to newer versions than 5.3.1.

Here's some CI jobs to validate the changes using your list:
- Linux: [Build Status](https://ci.ros2.org/job/ci_linux/14228/)
- macOS: [Build Status](https://ci.ros2.org/job/ci_osx/11919/)
- Windows: [Build Status](https://ci.ros2.org/job/ci_windows/14332/)

**Review by wjwwood — APPROVED** [review]

**Pull request merged**
